### ISSUE TITLE
Gradient of tf.reduce_sum(_, axis = _) throws an exception

### ISSUE THREAD

**Issue by mishun** [SciSharp/TensorFlow.NET#822]
Hi!

This example (it creates 2x2 matrix, uses tf.reduce_sum on it and tries to differentiate the result):
```F#
open Tensorflow
open type Tensorflow.Binding

[<EntryPoint>]
let main argv =
    tf.compat.v1.disable_eager_execution()
    let x = tf.placeholder(tf.float64, shape = TensorShape(2))
    let sq = tf.matmul(tf.reshape(x, shape = TensorShape(2, 1)), tf.reshape(x, shape = TensorShape(1, 2))) // Make 2z2 matrix
    let y = tf.reduce_sum(sq, axis = 0)
    let g = tf.gradients(y, x)
    begin
        use session = tf.Session()
        let res = session.run(g.[0], [| FeedItem(x, [| -1.0 ; 1.0 |]) |])
        printfn "result = %O" res
    end
    0
```
causes following exception:
```
Unhandled exception. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.NotImplementedException
   at Tensorflow.Gradients.math_grad._SumGrad(Operation op, Tensor[] grads)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.InvokeMember(String name, BindingFlags bindingFlags, Binder binder, Object target, Object[] providedArgs, ParameterModifier[] modifiers, CultureInfo culture, String[] namedParams)
   at Tensorflow.ops.<>c__DisplayClass1_1.<RegisterFromAssembly>b__3(Operation oper, Tensor[] out_grads)
   at Tensorflow.gradients_util._MaybeCompile(String scope, Operation op, Tensor[] out_grads, Action func, Func`3 grad_fn)
   at Tensorflow.gradients_util.<>c__DisplayClass0_2.<_GradientsHelper>b__6(NameScope scope1)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.gradients_util.<>c__DisplayClass0_0.<_GradientsHelper>b__0(NameScope scope)
   at Tensorflow.Binding.tf_with[T](T py, Action`1 action)
   at Tensorflow.gradients_util._GradientsHelper(Tensor[] ys, Tensor[] xs, Tensor[] grad_ys, String name, Boolean colocate_gradients_with_ops, Boolean gate_gradients, Int32 aggregation_method, Tensor[] stop_gradients, Graph src_graph)
   at Tensorflow.tensorflow.gradients(Tensor ys, Tensor xs, Tensor[] grad_ys, String name, Boolean colocate_gradients_with_ops, Boolean gate_gradients, Nullable`1 aggregation_method, Tensor[] stop_gradients)
   at Program.main(String[] argv) in C:\Users\user\source\repos\TF-Test\Program.fs:line 10
```

**Comment by mishun** [comment]
I tried to look into fixing that and, if I understand correctly, these lines are culprit:
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L530-L533

Now I don't understand how `_SumGrad` works. Isn't is supposed to look at `keep_dims` here:
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L514-L515
and here:
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L519-L520
? Why is -1 here:
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L530
? And why these things:
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L549
https://github.com/SciSharp/TensorFlow.NET/blob/6adcfaea3ff6f5c9d2437e54b66d781c4c16a72a/src/TensorFlowNET.Core/Gradients/math_grad.cs#L558
do something different from each other?

**Comment by Oceania2018** [comment]
Can you update with latest version?
If the issue still exist, please reopen this issue and PR a C# unit test to repreduce.

**Comment by mishun** [comment]
I haven't checked it yet, but since nothing has changed in
https://github.com/SciSharp/TensorFlow.NET/blob/6d2135833151adeb5aae5801256c8ee3b40ea652/src/TensorFlowNET.Core/Gradients/math_grad.cs#L497
for pretty long time I'm reasonably sure that it still doesn't work. Equivalent part in tensorflow's code suggests that there's some optimization there that just isn't implemented:
https://github.com/tensorflow/tensorflow/blob/21cc3eb8393991e312a2c2de397508cb297364d6/tensorflow/python/ops/math_grad.py#L149
On the other hand, just leaving non-optimized version is probably better than throwing an exception.

Could you remind me where unit tests for non-eager differentiation go? (I PR'ed a couple tests there a couple years ago, but forgot since)

**Comment by Oceania2018** [comment]
@mishun You can PR a test in [this file](https://github.com/SciSharp/TensorFlow.NET/blob/master/test/TensorFlowNET.Graph.UnitTest/GradientTest/GradientTest.cs).

**Comment by mishun** [comment]
Test case added: #917

**Comment by mishun** [comment]
Looks like I have no rights to reopen this one. Should I create new issue?
